### ISSUE TITLE
Abstract @Test method

### ISSUE THREAD

**Issue by bootstraponline** [linkedin/dex-test-parser#13]
I think it's a bug that this test is detected as valid.

```kotlin
public abstract class TeacherTest {

    @Test
    public abstract void displaysPageObjects();
}
```

> java.lang.InstantiationException: Can't instantiate abstract class com.instructure.teacher.ui.utils.TeacherTest

**Comment by drewhannay** [comment]
The subject says Kotlin, but the code snippet is Java code. It looks like this is an issue in both languages though. I'm not sure if there's any valid reason to have `@Test` on an abstract method in JUnit, but I'll do some research.

I probably won't have time to take a look at this until after Google I/O next week, but I'll follow up.

**Comment by bootstraponline** [comment]
I'm at I/O this week as well. We're using dex-test-parse to power Flank. It'd be cool to meet up if our schedules align. I'm on the firebase slack if you want to chat.

https://firebase.community

I worked around this issue by removing the annotation.

**Comment by kkoser** [comment]
I just did some testing on this in a test project, and it looks like adb instrument with the Junit4 runner DOES notice and run implementations of abstract methods that were annotated with @Test, even if the concrete implementation of that method does not. So it's probably also something dex-test-parser should support, although I'm having trouble thinking of a good reason to use this functionality...

**Comment by bootstraponline** [comment]
The issue is causing problems for Flank.
- https://github.com/TestArmada/flank/issues/253

For Firebase Test Lab, we need the ability to filter out the abstract classes.

**Comment by bootstraponline** [comment]
```
If I run this text locally or directly with Firebase (without using Flank) the list of tests executed will be:
espresso.fail.multidex.FooTest#testBar
espresso.fail.multidex.FooTest#testMainActivity
espresso.fail.multidex.FooTest#testAbstractBar
espresso.fail.multidex.FooTest#testAbstractBar2

but if I use Flank I will get:
espresso.fail.multidex.FooTest#testBar
espresso.fail.multidex.FooTest#testMainActivity
espresso.fail.multidex.AbstractFooTest#testAbstractBar
espresso.fail.multidex.AbstractFooTest#testAbstractBar2
```

that seems like a valid use case? I'm not sure there's an easy way to resolve methods in the same way.

**Comment by kkoser** [comment]
See #16 for a solution

**Comment by drewhannay** [comment]
Fixed by #16 

Note that this is only fixed for JUnit 4 tests, not JUnit 3. Since Google has deprecated the JUnit 3 base classes in the Android SDK and is now recommending that new tests should be written with the Android Testing Support Library (which is JUnit 4), it's extremely unlikely we'll fix this issue for JUnit 3 (it's not a very straightforward change at all, and would likely have a negative impact on the runtime performance of dex-test-parser).
